lib/pbio/drv/usb/usb_ev3.c

```diff
@@ -52,500 +28,823 @@
-    // on the bus.
+    // Reset the PHY
+    HWREG(CFGCHIP2_USBPHYCTRL) |= CFGCHIP2_RESET;
+    for (int i = 0; i < 50; i++) {__asm__ volatile ("");
+    }
+    HWREG(CFGCHIP2_USBPHYCTRL) &= ~CFGCHIP2_RESET;
+
+    // Set up the PHY and force it into device mode
+    HWREG(CFGCHIP2_USBPHYCTRL) =
+        (HWREG(CFGCHIP2_USBPHYCTRL) &
+            ~CFGCHIP2_OTGMODE &
+            ~CFGCHIP2_PHYPWRDN &            // Make sure PHY is on
+            ~CFGCHIP2_OTGPWRDN) |           // Make sure OTG subsystem is on
+        CFGCHIP2_FORCE_DEVICE |             // We only ever want device operation
+        CFGCHIP2_DATPOL |                   // Data lines are *not* inverted
+        CFGCHIP2_SESENDEN |                 // Enable various analog comparators
+        CFGCHIP2_VBDTCTEN;
+
+    HWREG(CFGCHIP2_USBPHYCTRL) =
+        (HWREG(CFGCHIP2_USBPHYCTRL) & ~CFGCHIP2_REFFREQ) |
+        CFGCHIP2_REFFREQ_24MHZ |            // Clock is 24 MHz
+        CFGCHIP2_USB2PHYCLKMUX;             // Clock comes from PLL
+
+    // Wait for PHY clocks to be ready
+    while (!(HWREG(CFGCHIP2_USBPHYCTRL) & CFGCHIP2_PHYCLKGD)) {
+    }
+
+    // Enable "PDR" mode for handling interrupts
     //
-    USBDCDCInit(0, (tUSBDCDCDevice *)&g_sCDCDevice);
+    // The datasheet doesn't clearly explain what this means,
+    // but what it appears TI has done is to wrap some custom interrupt and DMA
+    // logic around the Mentor Graphics core. The standard core registers
+    // thus now live at offset +0x400, and addresses below that pertain to the wrapper.
+    // This leaves some redundancy with how interrupts are set up, and this bit
+    // seems to enable accessing everything the TI way (more convenient) rather than
+    // the standard Mentor Graphics way (interrupt flags spread across more registers).
+    HWREG(USB_0_OTGBASE + USB_0_CTRL) &= ~(1 << 3);
+    HWREGH(USB0_BASE + USB_O_TXIE) = 0x1f;
+    HWREGH(USB0_BASE + USB_O_RXIE) = 0x1e;
+    HWREGB(USB0_BASE + USB_O_IE) = 0xff;
+
+    // Enable the interrupts we actually care about
+    HWREG(USB_0_OTGBASE + USB_0_INTR_MASK_SET) =
+        (1 << 18) |         // reset
+        (1 << 9) |          // EP 1 OUT
+        (1 << 1) |          // EP 1 IN
+        (1 << 0);           // EP 0 control pipe
+
+    // Clear all the interrupts once
+    HWREG(USB_0_OTGBASE + USB_0_INTR_SRC_CLEAR) = HWREG(USB_0_OTGBASE + USB_0_INTR_SRC);
+
+    // Hook up interrupt handler
+    IntRegister(SYS_INT_USB0, usb_device_intr);
+    IntChannelSet(SYS_INT_USB0, 2);
+    IntSystemEnable(SYS_INT_USB0);
+
+    // Finally signal a connection
+    USBDevConnect(USB0_BASE);
+
+    // We are basically done. USB is event-driven, and so we don't have to block boot.
+    pbio_os_process_start(&pbdrv_usb_ev3_process, pbdrv_usb_ev3_process_thread, NULL);
 }
 
 pbdrv_usb_bcd_t pbdrv_usb_get_bcd(void) {
-    return g_bUSBConfigured ? PBDRV_USB_BCD_STANDARD_DOWNSTREAM : PBDRV_USB_BCD_NONE;
+    // TODO: Reimplement this
```

Comment:
There isn't USB battery charging on EV3, so no TODO here. We probably don't even need to implement the function `PBSYS_CONFIG_BATTERY_CHARGER` should not be defined (or 0) on EV3, so this function should never be called.